### PR TITLE
Pm/zap token to pt

### PR DESCRIPTION
**Pull request by Padraic-O-Mhuiris**

### CODE REVIEW

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (satyamakgec) [inline contracts/zaps/ZapCurveTokenToPrincipalToken.sol:358]: Can we put an example in the comments of how it gonna work? Like what will be the root token, curvePool , lpToken etc .. It will be better to understand.

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (satyamakgec) [inline contracts/zaps/ZapCurveTokenToPrincipalToken.sol:307]: How we are going to validate the index of the token that the user is going to get.

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (satyamakgec) [inline contracts/zaps/ZapCurveTokenToPrincipalToken.sol:370]: Again use the differential amount here instead of the balance directly.

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (satyamakgec) [inline contracts/zaps/ZapCurveTokenToPrincipalToken.sol:432]: Should not be zero

**Review by satyamakgec — CHANGES_REQUESTED** [review]
It is a preliminary review, I need to do another pass

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:221]: So the `_ctx` array is necessary as it is meant to be used to keep a context of what occurred in the `childZap` from the `zapIn()` function which called this.

As you can see in the excerpt below, for the `_childZap` it's irrelevant and we just pass an undeclared `uint256[3]`, however the result of the `_childZap` is inserted at the index relevant to the `_zap` amount. 

https://github.com/element-fi/elf-contracts/blob/pm/zapTokenToPt/contracts/zaps/ZapCurveTokenToPrincipalToken.sol#L273-L290

As an example, we have a principal token based on the curve pool token `LUSD3CRV_f` which is a metapool(?) that combines `LUSD` and the `3CRV` token. The `3CRV` token is based on `DAI`, `USDT` and `USDC`. Contextualising this, the user might only have `DAI` to purchase the PT with and so they would initially `_zapCurveLpIn` on the `_childZap` which contains the amount of `DAI` they have plus other info. As a result there would be an amount of `3CRV` tokens that would be recorded as a result in the `ctx` array, which is then passed to the `_zapCurveLpIn` on the `_zap` which would result in receiving `LUSD3CRV_f` tokens which are subsequently used to purchase the PT's.

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:190]: Yes that is much cleaner

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:239]: Yes that can be done. Might however introduce some UX issues as there is not an absolute deterministic solution for calculating the curve swap correctly (`calc_token_amount` is intended as loosely accurate -> https://github.com/curvefi/curve-contract/blob/master/contracts/pool-templates/a/SwapTemplateA.vy#L347-L355) but can leave it up to the frontend to determine that issue

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:244]: The reason for this was that in some cases, I think in particular the `STETH` token, there is a precision loss of 1 wei. If by the differential value, you mean the `zap.amounts[i]` figure, that would cause a revert in the curve call

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:281]: We could use the `coins` method on the curve contract maybe? It takes an integer value returning the address of the erc20 at that index. We could validate if the returned address from that call matches the childZap LP token? But we also would run into the issue of different interfaces for the `coins` method so that would require passing func sig for that from off-chain.

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:305]: Could you clarify this? We have the `setApprovalsFor` function which would enable the admin to do so?

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:369]: Added comment, LMK if ok

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:316]: We could do it by the `coins` method on the curve pool I mentioned in another comment above but may be more trouble than it's worth if func sig must be passed from off-chain

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:381]: Could you clarify exactly by what you mean as the "differential" amount?

**Review by NicholasDotSol — APPROVED** [review]
looks good

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol]: Yes but the only issue with doing so would be where someone could set it to approval to 0 just to be chaotic. It could be possible to have an open approve if the amount they wish to set the contracts approval to is greater than the current

Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol]: It actually would be best to leave as authorized as it requires consideration of allowance behaviour across all tokens we would support.

File: contracts/zaps/ZapCurveTokenToPrincipalToken.sol
Comment (Padraic-O-Mhuiris) [reply contracts/zaps/ZapCurveTokenToPrincipalToken.sol:90]: I think I've restructured it correctly as per your description @satyamakgec so am as well leave it maybe? But LMK either way

**Review by satyamakgec — APPROVED** [review]

**Pull request merged**
